core/versions_2.0.md

```diff
@@ -0,0 +1,71 @@
+
+If the workflow code is executing for the first time, it will write a marker into history meaning "change change-id happened". If it is replaying history, it will verify such a marker exists and return true. If the marker does not exist, it will return false. Versions of the code without the call will fail (more on that in other section).
+
+Version ranges can still be simulated if needed like:
+```rust
+if changed("change-ver-2") {
+    // blah
+} else if changed("change-ver-1") {
+    // blah
+}
+// etc
+```
+
+This seems dramatically more understandable at the cost of making a probably not very common scenario mildly more verbose.
+
+## Enforcing the presence of version API calls/commands in the presence of version markers (2-phase deprecation)
+To expand on why the second change is important. Imagine the below two versions of the same (imaginary, Rust) workflow code:
+
+```rust
+// Original version
+fn workflow() {
+   // ...previous code....
+   activity_a().await;
+   activity_must_run_after_a().await;
+}
+```
+
+```rust
+// Version 2
+fn workflow() {
+   // ...previous code....
+   if changed("change_id") {
+       activity_b().await;
+   } else {
+       activity_a().await;
+       activity_must_run_after_a().await;
+   }
+}
+```
+
+Imagine that a run of the workflow begins executing on a worker with version 2. It hits the `get_version` call and writes the marker, and then dies. Then, a worker running the original code picks up that history. It will execute the `a` activities, but it should not have, it was already known that this workflow history should take path B.
+
+With this change, the workflow task would fail because the worker's code does not have a `get_version` call, but there was a version marker in the history.
+
+This has the effect that users now must take a 2-phase approach to deprecating old code, but in exchange are given better guarantees around workflow state consistency. If the user wants to remove path A, they can change their code to:
+
+```rust
+// Version 3
+fn workflow() {
+   // ...previous code....
+   must_have_change("change_id")
+   activity_b().await;
+}
+```
+
+The above code can now fail properly if it encounters any histories which lack the version marker (were run by the old code), while still allowing the user to eliminate the old code. This new code does not write a marker if it is creating this part of history for the first time, rather than replaying.
+
+This means the user cannot deploy this "phase 2" code until they know there are no more workers out there running the original code (as, if such a worker were to encounter a history generated by this latest code, it would not see any marker and proceed). Hence, it is important to provide a way for users to query workers so they can know this (apparently, already exists with enhanced visibility?). After no more workflow runs exist which still contain the marker, the `must_have_change` call may be removed entirely.
```

Comment:
Using a different marker as suggested in my comment above means that it's safe to deploy version 3 alongside version 2.
The only thing stopping you from deploying version 3 is WF histories without the marker (produced by version 1), it's still "safe" but those WFs cannot be processed and their tasks will fail.

Reply:
I think you mean safe to deploy 3 alongside 1, (in my original description 3 w/ 2 is still safe) but yeah good point.